epsilon_transformers/process/MixedStatePresentation.py

```diff
@@ -36,7 +41,29 @@ def paths(self) -> List[List[int]]:
 	
 	@property
 	def paths_and_belief_states(self) -> List[Tuple[List[int], Float[np.ndarray, "n_states"]]]:
-		return [(x.path, x.state_prob_vector) for x in self.nodes]
+		paths_and_beliefs = [(x.path, x.state_prob_vector) for x in self.nodes]
+		paths = [x[0] for x in paths_and_beliefs]
+		beliefs = [x[1] for x in paths_and_beliefs]
+		return paths, beliefs
+	
+	@property
+	def block_entropy(self) -> Float[np.ndarray, "depth"]:
+		depth_emission_probs = [[] for _ in range(self.depth)]
+		
+		def traverse(node: MixedStateTreeNode, depth: int, accumulated_prob: float):
+			if depth < self.depth:
+				if node is not self.root_node:
+					depth_emission_probs[depth].append(accumulated_prob * node.emission_prob)
+				for child in node.children:
+					traverse(child, depth + 1, accumulated_prob * node.emission_prob if node is not self.root_node else 1.0)
+		
+		traverse(self.root_node, 0, 1.0)
```

Comment:
make traverse pure. 

move `depth_emission_probs = [[] for _ in range(self.depth)]` into traverse and have it return from traverse. Include in type signature. when traverse is called, save the output.

move traverse into _traverse

Reply:
changed code to

`def get_beliefs_for_transformer_inputs(
    transformer_inputs: Float[torch.Tensor, "batch n_ctx"], 
    msp_belief_index: Dict[Tuple[float, ...], int], 
    tree_paths: List[List[int]], 
    tree_beliefs: List[List[float]]
) -> Tuple[Float[torch.Tensor, "batch n_ctx belief_dim"], Float[torch.Tensor, "batch n_ctx"]]:
    batch, n_ctx = transformer_inputs.shape
    belief_dim = len(list(msp_belief_index.keys())[0])
    path_belief_dict = {tuple(path): belief for path, belief in zip(tree_paths, tree_beliefs)}
    X_beliefs = torch.zeros(batch, n_ctx, belief_dim, device=transformer_inputs.device)
    X_belief_indices = torch.zeros(batch, n_ctx, dtype=torch.int, device=transformer_inputs.device)`

Reply:
```python
	@property
	def block_entropy(self) -> Float[np.ndarray, "depth"]:
		depth_emission_probs = self._traverse(self.root_node, 0, 1.0, depth_emission_probs)
		block_entropy = np.array([entropy(probs) if probs else 0 for probs in depth_emission_probs])
		return block_entropy

	def _traverse(self, node: MixedStateTreeNode, depth: int, accumulated_prob: float, depth_emission_probs):
		depth_emission_probs = [[] for _ in range(self.depth)]
		if depth < self.depth:
			if node is not self.root_node:
				depth_emission_probs[depth].append(accumulated_prob * node.emission_prob)
			for child in node.children:
				depth_emission_probs = self._traverse(child, depth + 1, accumulated_prob * node.emission_prob if node is not self.root_node else 1.0, depth_emission_probs)
		return depth_emission_probs
```